rfcs/0014-abi-errors.md

```diff
@@ -0,0 +1,456 @@
+- Feature Name: `error_type`
+- Start Date: 2024-11-18
+- RFC PR: [FuelLabs/sway-rfcs#43](https://github.com/FuelLabs/sway-rfcs/pull/43)
+- Sway Issue: [FueLabs/sway#0000](https://github.com/FuelLabs/sway/issues/001)
+
+# Summary
+
+[summary]: #summary
+
+This RFC aims to standardize error handling in Sway so that there is a
+deliberate and clear way to produce error messages that are easily consumable by
+language tools while keeping error strings out of on-chain artifacts.
+
+# Motivation
+
+[motivation]: #motivation
+
+Currently, reverting a sway program only produces an opaque error code.
+Sometimes this code is unusual enough that with the source code of the contract
+it is possible to figure out what is going on, but a lot of the time it's a
+meaningless number that only means that some failure happened somewhere.
+
+It's not reasonable to expect people to be able to debug sway programs in these
+conditions, so we need to improve this.
+
+We want irrecoverable errors to carry user defined error messages and source
+information about where the error was invoked. We want those error types to be
+something that can also be used for recoverable errors.
+
+To this end we want to introduce a mechanism that makes error codes an
+understandable part of the ABI, and language features that allow the user to
+define error messages, error types and invoke those errors with the appropriate
+information being generated.
+
+We also want recoverable errors to connect to this mechanism so that all error
+handling in a given Sway program can go through one or a small set of user
+defined types and error strings don't affect the final bytecode size.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+## Error types
+
+Error types are special enums that allow developers to signal that each variant
+is associated with an error message.
+
+An error type implements the `Error` trait, but it is not recommended that you
```

Comment:
Ultimately the issue is that we need some way of telling if a type is an error and to enforce that any such type:
* is an enum
* has an error message annotation for all of its variants

I guess we could do away with the marker and just have this be enforced by the impl of the marker type? I'm not sure if that's more intuitive or not.

@tritao I'm curious what you think here.

Reply:
## TL;DR;
- We need marker traits in general.
- We can unify input to `panic` via marker trait.
- I am not sure if we can have a useful `Error` trait that can actually have implementation.

## The need for marker traits in general

As mentioned, we will need marker traits in general. E.g., in order to express the needed concepts explicitly, without "magic", the Storage 3.0 RFC calls for (work title) [Serializable marker trait](https://github.com/FuelLabs/sway-rfcs/blob/7a1cb3fd429c79e4c318f7675e04fb479d63bf90/files/0013-configurable-and-composable-storage/sway-libs/storage/storage_box.sw#L6).

Also, note that we already assign properties to types, but unlike explicitly by assigning them marker traits, we analyze them on usage site within the compiler and check if they satisfy certain properties. E.g., `AbiEncode` cannot be auto-implemented for types that have pointers. The above link to the proposed (work title) `Serializable` marker trait explains how that trait represent the property, among others, of not having pointers. Having that marker trait would lead to the rule stated explicitly in terms of the type system. E.g., `AbiEncode` is auto-implemented for all types that are marked with `Serializable` marker trait.

This are just some examples why I see the need for marker traits in general.

So yes, assigning them to types is a part of the semantic analysis. When analyzing a type we mark it with all the appropriate marker traits.

In the case of enums that are errors, there is no heuristics in this decision. As per RFC, only those enums that are attributed with `#[error_type]` are marked with the `Error` marker trait.

I see here more issues in naming details, like, shouldn't it be then `core::marker::Error`, etc. But all these details and conventions can be defined when defining the general approach to marker traits in Sway.

The most important feature of Sway marker traits I see, is, that unlike a few of Rust marker traits, they can never be implemented by developers. They are always exclusively assigned by the compiler. (In Rust, for some, it is possible to `unsafe` implement them.)

## Unifying input to `panic` via marker trait `Error`

If we go for this approach, we would define the `Error` marker trait, as: any type that can be accepted by the `panic`. Note that this does not exclude it from being use in `Result`.

An interesting consequence of this is, that `str` can also be marked by the `Error` marker. This then simply means, we can pass `str` to `panic`. There is no special rule for `str` in regards to `panic`. This actually makes `panic` unified. Strictly speaking, the remark that @maschad had about mixing `str` and enums wouldn't be there any more. `panic` accepts `T: Error` and`str` is also marked with the `Error` trait. No exceptions.

## Is a non-marker `Error` actually possible?

Above I've sketched something like:

```
const trait Error {
    fn description(&self) -> str;
}
```

This makes sense at the first look, but in practice, the `Error` instance passed to `panic` must not and will not be a constant. Meaning, in a general/regular case we cannot const evaluate the `description()` as defined above. E.g., to expand on the example from the RFC, all of this will work as per RFC and if the `Error` is just an empty marker trait:

```
pub fn main(a: u8, b: u64) {
  match a {
    4 => panic(MyError::B((b, a))),
    5 => panic(MyError::B((a.as_u64(), b.as_u8()))),
    _ => panic(some_non_const_evaluation_of_my_error(a, b))),
  }
}
```
Note that none of the inputs to `panic`s above is const-evaluable. And it doesn't need to be. For the proposal of the RFC to work, to generate the expected ABI fields, _the compiler only needs to know the type of the provided `Error` instance_. The wiring of the logged instance to the actual error message is done by the tooling later on, as explained in the RFC.

In other words, I don't see how the sketched `const trait Error` implemented by a developer would work in the above example.

The more I think about it, the more I see the approach proposed in the RFC as the right one. The error messages are fixed compile time constants attached to enum variants, not to their particular instances. In the second step, we will allow them to be format strings that can reffere to the logged instance parts, but the calculation of the actual string will be done by the tooling, and not by the compiler.

I hope this all explains why I am for defining marker traits in general and starting here with the marker trait. And even perhaps staying on it long term :wink:

Reply:
I think this is an interesting discussion but is ultimately off topic since it's an implementation detail.

I agree with most of your points in the utility of having marker traits (even marked as such as a specific concept) but I don't think we should burden the spec with those details.

Suffice to say, a trait should exist so that we can check the type correctness of `panic` invocations, and this is the most expedient solution. Future evolutions of the trait to eliminate some of the magical syntax are, well, future considerations.